parse.go

```diff
@@ -0,0 +1,395 @@
+package mf2
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+)
+
+type parser struct {
+	lexer *lexer
+	items []item
+	pos   int
+}
+
+func (p *parser) next() item {
+	if p.pos >= len(p.items) {
+		// todo: error or check elsewhere
+		return item{}
+	}
+
+	p.pos++
+	return p.items[p.pos]
+}
+
+func (p *parser) current() item {
+	return p.items[p.pos]
+}
+
+func (p *parser) collect() error {
+	// sanity check, avoid infinite loop
+	for i := 0; i < 1000; i++ {
+		item := p.lexer.nextItem()
+		if item.typ == itemError {
+			return fmt.Errorf("got error token: %s", item.val)
+		}
+
+		p.items = append(p.items, item)
+
+		if item.typ == itemEOF {
+			return nil
+		}
+
+	}
+
+	return errors.New("too many tokens. infinite loop ?")
+}
+
+func new(lexer *lexer) *parser {
```

Comment:
We should not override system-predefined identifiers. This configures `predefined` lint rule https://github.com/expect-digital/go-mf2/pull/6